scipy/integrate/_tanhsinh.py

```diff
@@ -691,7 +690,7 @@ def _euler_maclaurin_sum(fj, work, xp):
     fjwj = fj + xp.log(work.wj) if work.log else fj * work.wj
 
     # update integral estimate
-    Sn = (_logsumexp(fjwj + xp.log(work.h), axis=-1, xp=xp) if work.log
+    Sn = (special.logsumexp(fjwj + xp.log(work.h), axis=-1) if work.log
```

Comment:
Since Lucas can't reproduce the failures locally, I'll just note that this line seems to be the key to the two test failures in CI which I can also reproduce locally with torch `2.4.0`.

The tests pass with a partial restoration here (ignoring the boilerplate parts of the restoration diff):

```diff
     # update integral estimate
-    Sn = (special.logsumexp(fjwj + xp.log(work.h), axis=-1) if work.log
+    Sn = (_logsumexp(fjwj + xp.log(work.h), axis=-1, xp=xp) if work.log
           else xp.sum(fjwj, axis=-1) * work.h)
```

If I prefix with `SCIPY_DEVICE=cuda` those two test cases actually pass (not skip) on the current version of this branch. Have we added a "gpu only" run mode for tests with backends yet? I'm sure folks would appreciate even more complexity in the test harnesses.

The NumPy version in the environment appears to be irrelevant within our supported range for the `torch` failures. That's probably as deep as I'll get on this one this evening, but it is a new failure vs. `main`.

Reply:
Hmm, strange, I still can't reproduce the failures with torch `2.4.0`:

```
(Pdb) xp.exp(res.integral)
tensor(1.4189-3.4754e-16j, dtype=torch.complex128)
(Pdb) ref.integral
tensor(1.4189, dtype=torch.float64)
...
(Pdb) xp.exp(res.integral)
tensor(1.4189-3.4754e-16j, dtype=torch.complex128)
(Pdb) ref.integral
tensor(1.4189, dtype=torch.float64)
```

> I'm sure folks would appreciate even more complexity in the test harnesses.

😁

Reply:
> Have we added a "gpu only" run mode for tests with backends yet?

On a more serious note, I don't think we would want to add this, although it can be simulated with

```python
@skip_xp_backends(
    np_only=True,
    exceptions=['list', 'of', 'gpu', 'backends'],
)
```

Reply:
> failures in CI which I can also reproduce locally with torch 2.4.0.

could you provide the command and environment you used to reproduce? On my MacBook, with either torch 2.3 or torch 2.4, and either NumPy 2.0 or 2.1, `python dev.py test -b pytorch -s integrate` passes. The failing `test_log` tests are ran and pass.